src/stk/math/types.h

```diff
@@ -131,50 +132,97 @@ struct STK_ALIGN(16) double4
 #endif // STK_USE_CUDA
 
 
-template<typename T, unsigned int rows_, unsigned int cols_>
-struct matrix
+struct Matrix3x3f
 {
-    T _data[rows_][cols_];
+    static constexpr unsigned int rows = 3;
+    static constexpr unsigned int cols = 3;
 
-    static constexpr unsigned int rows = rows_;
-    static constexpr unsigned int cols = cols_;
+    float3 _rows[rows];
 
-    T* operator[](const unsigned int i) {
-        return _data[i];
+    float3& operator[](const unsigned int i) {
+        return _rows[i];
     }
 
-    const T* operator[](const unsigned int i) const {
-        return _data[i];
+    const float3& operator[](const unsigned int i) const {
+        return _rows[i];
     }
 
-    T& operator()(const unsigned int r, const unsigned int c) {
-        return _data[r][c];
+    const float& operator()(const unsigned int r, const unsigned int c) const {
```

Comment:
This is for convenience, but it should not be a performance-critical component.

Reply:
This was a bit odd, maybe something like this instead:

```
ASSERT(c < 3 && r < 3);
return *reinterpret_cast<float*>(&_rows[r])+c;
```

Reply:
I have this resolution to do things The C++ Way™, trying to avoid unsafe conversions, but I must admit the C++ way usually sucks, so I will go for your suggestion :-D